native/common/jp_classhints.cpp

```diff
@@ -56,17 +56,21 @@ jvalue JPMatch::convert()
 	return conversion->convert(*this);
 }
 
-JPMethodMatch::JPMethodMatch(JPJavaFrame &frame, JPPyObjectVector& args)
-: argument(args.size())
-{
-	type = JPMatch::_none;
-	isVarIndirect = false;
-	overload = 0;
-	offset = 0;
-	skip = 0;
+JPMethodMatch::JPMethodMatch(JPJavaFrame &frame, JPPyObjectVector& args, bool callInstance)
+: m_Arguments(args.size())
+{
+	m_Type = JPMatch::_none;
+	m_IsVarIndirect = false;
+	m_Overload = 0;
+	m_Offset = 0;
+	m_Skip = 0;
+	m_Cache = callInstance ? 0 : 1000;
 	for (size_t i = 0; i < args.size(); ++i)
 	{
-		argument[i] = JPMatch(&frame, args[i]);
+		PyObject *arg = args[i];
+		m_Arguments[i] = JPMatch(&frame, arg);
+		m_Cache *= 0x10523C01;
+		m_Cache += (long) (Py_TYPE(arg));
```

Comment:
This assumes that all casts of the pointer ob_type are smaller than the magic number to achieve proper separation of all types, am I wrong? I like this idea, but can you demonstrate that this holds in most or all cases?

Reply:
See comment above.   If you are worried I can actually generate a real LCG magic number (assuming c is zero) by creating a number which is prime in a field of 2^64.  It is just more effort.